runtime/oti/VMHelpers.hpp

```diff
@@ -998,6 +998,190 @@ class VM_VMHelpers
+	{
+		bool isASCII = true;
+
+		for (UDATA i = 0; i < length; ++i) {
+			if (data[i] > 0x7F) {
+				isASCII = false;
+				break;
+			}
+		}
+
+		return isASCII;
+	}
+
+	/**
+	 * Copies a UTF8 string into a UTF16 string at a specific index.
+	 *
+	 * @param vmThread[in] the current J9VMThread
+	 * @param data[in] points to raw UTF8 bytes, some of which may be multi-byte UTF8 encoded Unicode characters
+	 * @param length[in] the number of bytes representing characters in data
+	 * @param stringFlags[in] string flags corresponding to data
+	 * @param charArray[in] the character array (can be either byte[] or char[]) to copy the UTF8 string into
+	 * @param startIndex the start index of charArray at which to begin the copy
+	 */
+	static VMINLINE void
+	copyUTF8ToUTF16(J9VMThread * vmThread, U_8 * data, UDATA length, UDATA stringFlags, j9object_t charArray, UDATA startIndex)
+	{
+		UDATA originalLength = length;
+
+		if (J9_ARE_ALL_BITS_SET(stringFlags, J9_STR_ASCII)) {
+			if (J9_ARE_ANY_BITS_SET(stringFlags, J9_STR_XLAT)) {
+				for (UDATA i = startIndex; i < length; ++i) {
+					U_8 c = data[i];
+					J9JAVAARRAYOFCHAR_STORE(vmThread, charArray, i, c != '/' ? c : '.');
+				}
+			} else {
+				for (UDATA i = startIndex; i < length; ++i) {
+					J9JAVAARRAYOFCHAR_STORE(vmThread, charArray, i, data[i]);
+				}
+			}
+		} else {
+			UDATA writeIndex = startIndex;
+			while (length > 0) {
+				U_16 unicode = 0;
+				UDATA consumed = decodeUTF8Char(data, &unicode);
+				if (J9_ARE_ANY_BITS_SET(stringFlags, J9_STR_XLAT)) {
+					if ((U_16)'/' == unicode) {
+						unicode = (U_16)'.';
+					}
+				}
+				J9JAVAARRAYOFCHAR_STORE(vmThread, charArray, writeIndex, unicode);
+				writeIndex += 1;
+				data += consumed;
+				length -= consumed;
+			}
+		}
+
+		/* Anonymous classes have the following name format [className]/[ROMADDRESS], so we have to to fix up the name
+		 * because the previous loops have converted '/' to '.' already.
+		 */
+		if (J9_ARE_ALL_BITS_SET(stringFlags, J9_STR_ANON_CLASS_NAME)) {
+			for (IDATA i = (IDATA) originalLength - 1; i >= 0; --i) {
+				if ((U_16)'.' == J9JAVAARRAYOFCHAR_LOAD(vmThread, charArray, i)) {
+					J9JAVAARRAYOFCHAR_STORE(vmThread, charArray, i, (U_16)'/');
+					break;
+				}
+			}
+		}
+	}
+
+	/**
+	 * Copies a UTF8 string into a ASCII compact string at a specific index.
+	 *
+	 * @param vmThread[in] the current J9VMThread
+	 * @param data[in] points to raw UTF8 bytes, all of which are within the ASCII subset ord. [0, 127]
+	 * @param length[in] the number of bytes representing characters in data
+	 * @param stringFlags[in] string flags corresponding to data
+	 * @param charArray[in] the character array (can be either byte[] or char[]) to copy the UTF8 string into
+	 * @param startIndex the start index of charArray at which to begin the copy
+	 */
+	static VMINLINE void
+	copyUTF8ToASCII(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA stringFlags, j9object_t charArray, UDATA startIndex)
+	{
```

Comment:
Should there be a Trc_Assert that the `charArray` is of type `byte[]`?

Reply:
This is not true in general. It is a `byte[]` in JDK11, but a `char[]` in JDK8. The API does not really care what the type of the array is. All it cares is that we have enough space to put the ASCII characters inside the array, whatever it may be. Similarly to above should we change the name to not use `byte` or `char` in the name of this parameter to avoid confusion? Suggestions welcome on a better name. Best one I can think of at the moment is `backingArray`